Implement partial verification for CMS with multiple signatures

jluebbe opened the pull request:
Currently, OpenSSL supports creating CMS SignedData structures with
multiple signers using `CMS_sign()` with the `CMS_PARTIAL` flag combined
with `CMS_add1_signer()`. This can be done for a new SignedData or by
adding signers to an existing structure. In addition to the C API, this
is supported via `openssl cms -sign` (using multiple `-signer` options) and
`openssl cms -resign`.

The corresponding `CMS_verify()` function and `openssl cms -verify` command
check the signatures sequentially and abort with an error on the first
problem. They support some flags (`CMS_NO_SIGNER_CERT_VERIFY`,
`CMS_NO_ATTR_VERIFY`, `CMS_NO_CONTENT_VERIFY`), which disable some aspects
of the verification, but do not allow continuing after a failed
signature verification.

The main use cases for multiple signatures on one message are:

* Simplify CA certificate roll-over by signing messages with keys in
  both the old and new hierarchies during a transition period.
  Recipients would have either the old or new CA installed (or
  temporarily even both). In all cases, we'd like to accept any message
  which has at least one signature that can be verified.

* Support user-implemented verification policies which require
  signatures from multiple specific signers.

* Support user-implemented verification policies which require a minimum
  number of signatures from different signers under under a trusted CA.

This is made possible by adding a `CMS_VERIFY_PARTIAL` flag to `CMS_verify()`
and exposing it through a new `-verify_partial` option to `openssl cms
-verify`.

If this flag is set, the call is successful even if some of the
individual signatures cannot be verified (perhaps due to CAs missing
from the local store or expired certificates). The application would
then call `CMS_get0_signers()` and check if the set of valid signatures
satisfies its policy.

When verifying CMS messages with multiple signatures using
`CMS_VERIFY_PARTIAL` to implement additional custom requirements
regarding signatures from which certificates are required and/or
acceptable, information on which signature was considered invalid and
why by `CMS_verify` is useful for reporting or debugging.

Implement this by adding `CMS_SignerInfo`-based access to certificates
(`CMS_SignerInfo_get0_signer_cert`) and verification results
(`CMS_SignerInfo_get_verification_result`). Also use these new functions
to show more details when using `openssl cms -verify` with
`-verify_partial`.

Fixes #3028
Fixes #26382

##### Checklist
- [x] documentation is added or updated
- [x] tests are added or updated

jluebbe committed:
Implement partial verification for CMS with multiple signatures

Currently, OpenSSL supports creating CMS SignedData structures with
multiple signers using CMS_sign() with the CMS_PARTIAL flag combined
with CMS_add1_signer(). This can be done for a new SignedData or by
adding signers to an existing structure. In addition to the C API, this
is supported via openssl cms -sign (using multiple -signer options) and
openssl cms -resign.

The corresponding CMS_verify() function and openssl cms -verify command
check the signatures sequentially and abort with an error on the first
problem. They support some flags (CMS_NO_SIGNER_CERT_VERIFY,
CMS_NO_ATTR_VERIFY, CMS_NO_CONTENT_VERIFY), which disable some aspects
of the verification, but do not allow continuing after a failed
signature verification.

The main use cases for multiple signatures on one message are:

* Simplify CA certificate roll-over by signing messages with keys in
  both the old and new hierarchies during a transition period.
  Recipients would have either the old or new CA installed (or
  temporarily even both). In all cases, we'd like to accept any message
  which as at least one signature that can be verified.

* Support user-implemented verification policies which require
  signatures from multiple specific signers.

* Support user-implemented verification policies which require a minimum
  number of signatures from different signers under under a trusted CA.

This is made possible by adding a CMS_VERIFY_PARTIAL flag to CMS_verify()
and exposing it through a new -verify_partial option to 'openssl cms
-verify'.

If this flag is set, the call is successful even if some of the
individual signatures cannot be verified (perhaps due to CAs missing
from the local store or expired certificates). The application would
then call CMS_get0_signers() and check if the set of valid signatures
satisfies its policy.

Fixes #3028
Fixes #26382

jluebbe committed:
Implement access to CMS verification results per signer

When verifying CMS messages with multiple signatures using
CMS_VERIFY_PARTIAL to implement additional custom requirements
regarding signatures from which certificates are required and/or
acceptable, information on which signature was considered invalid and
why by CMS_verify is useful for reporting or debugging.

Implement this by adding CMS_SignerInfo-based access to certificates
(CMS_SignerInfo_get0_signer_cert) and verification results
(CMS_SignerInfo_get_verification_result). Also use these new functions
to show more details when using 'openssl cms -verify' with
'-verify_partial'.

jluebbe committed:
Update CHANGES.md for CMS_VERIFY_PARTIAL

jluebbe commented:
I'm not sure how to fix https://github.com/openssl/openssl/actions/runs/14977754186/job/42074263523?pr=27604#step:3:32, as the surrounding defines have the same indent.

t8m commented:
> I'm not sure how to fix https://github.com/openssl/openssl/actions/runs/14977754186/job/42074263523?pr=27604#step:3:32, as the surrounding defines have the same indent.

That can be waived.

shahsb reviewed (commented):
Thanks @jluebbe, for making the changes.

I have few design questions:

1) How do you prevent applications from misinterpreting partial verification as a guarantee of message integrity when only one weak/compromised signature is valid?
2) If an attacker adds a valid signature from a low-security CA, could this bypass stricter policies? The approach shifts policy enforcement to applications -- how do you ensure developers don’t overlook critical checks (e.g., requiring specific CAs)?
3) How are certificate revocation (CRL/OCSP) and timestamping handled in per-signer results?
4) A signature might be valid cryptographically but revoked. Are we considering the revocation status checks (CRL/OCSP) as part of this? If not, should we? How do applications enforce revocation-aware policies?
5) Should we consider adding utility functions (e.g., CMS_verify_policy_min_signatures()) to enforce common policies (minimum valid signatures, required CAs) to reduce developer error?

jluebbe commented:
Thanks for the questions, @shahsb!

> I have few design questions:

> 1. How do you prevent applications from misinterpreting partial verification as a guarantee of message integrity when only one weak/compromised signature is valid?

One needs to keep in mind that each CMS signature is independent of the others in the ASN.1 structure. This means that an attacker can easily remove any signature they want and pass on a subset.

So, if a weak signature exists (or can be produced) that is acceptable to a given configuration of `CMS_verify`, the attacker would just remove all other signatures and be able to satisfy the checks in the existing implementation.

Said differently, the current requirement that **all** signature can be verified doesn't really increase security, as the attacker could just remove any signatures that cause the verification to fail.

> 2. If an attacker adds a valid signature from a low-security CA, could this bypass stricter policies? The approach shifts policy enforcement to applications -- how do you ensure developers don’t overlook critical checks (e.g., requiring specific CAs)?

Keeping in mind that the attacker can also remove signatures, applications already need to make sure that only "strong enough" CAs are in the X509_STORE and no weak algorithms are allowed.

The intention of this PR is that:
* adding the `CMS_VERIFY_PARTIAL` flag should make no difference for a CMS with a single signature
* overall verification should be successful when at least one individual signature would be considered valid after splitting one CMS with N signatures into N CMSs with one signature each (which basically is the workaround suggested in https://github.com/openssl/openssl/issues/3028#issuecomment-2361380233 and could also be done by an attacker with the current implementation)

I should probably add more guidance to the manual pages on what to consider when checking each signer separately. But as long as the developer doesn't ignore the return value of `CMS_verify`, using `CMS_VERIFY_PARTIAL` wouldn't be less secure.

> 3. How are certificate revocation (CRL/OCSP) and timestamping handled in per-signer results?
> 4. A signature might be valid cryptographically but revoked. Are we considering the revocation status checks (CRL/OCSP) as part of this? If not, should we? How do applications enforce revocation-aware policies?

Responding to both questions together.

These are already configurable using `CMS_verify` flags (e.g. `CMS_NOCRL`) or `X509_VERIFY_PARAM_set_time()` to modify the validity checking behavior.
There should be no change by this PR to this aspect.

> 5. Should we consider adding utility functions (e.g., CMS_verify_policy_min_signatures()) to enforce common policies (minimum valid signatures, required CAs) to reduce developer error?

At least for our use cases, checking only the number of signatures (as `CMS_verify_policy_min_signature()` might do), wouldn't be very useful, as we'd be looking for signers with a specific combination of properties.

One specific check, for which a helper could be useful, could be rejecting a CMS with multiple signatures by the same certificate. That could be a follow-up PR, tough.

openssl-machine commented:
This PR is in a state where it requires action by @openssl/committers but the last update was 30 days ago

openssl-machine commented:
This PR is in a state where it requires action by @openssl/committers but the last update was 61 days ago

DDvO commented:
Hint: by placing additions to global lists such as `CHANGES.md` and `util/libcrypto.num` not at their very end where everyone else also keeps adding their new pieces, but somewhere in between, strongly reduces chances of getting trivial but tedious merge conflicts.

jluebbe commented:
@DDvO I've rebased to solve the conflicts and moved the additions to CHANGES.md and libcrypto.num to other locations. Otherwise the PR is unchanged. Thanks for the hint.

jluebbe commented:
I've rebased to solve a conflict in `test/recipes/80-test_cms.t`.

openssl-machine commented:
This PR is in a state where it requires action by @openssl/committers but the last update was 30 days ago

dndfrink-ibm commented:
@jluebbe Are you still planning on merging this commit? I would be interested in using this functionality if it were to exist

jluebbe commented:
> @jluebbe Are you still planning on merging this commit? I would be interested in using this functionality if it were to exist

Yes, definitely. I got some feedback from @shahsb, which I think I've addressed. I assume the next step would be a review by a maintainer.

What is your use-case for this?

I'll rebase again to solve the conflicts and make it easier for you to test.

dndfrink-ibm commented:
Thanks for the response/rebase! Vaguely, I want to use CMS for secure boot verification where only a subset of signatures in the CMS structure need to be verified for the operation to succeed. I think it's possible to do this without these changes but this makes it much easier for me to implement.

DDvO commented on apps/cms.c:
```diff
@@ -1333,6 +1338,39 @@ int cms_main(int argc, char **argv)
                 ret = verify_err + 32;
             goto end;
         }
+        if (flags & CMS_VERIFY_PARTIAL) {
```
```suggestion
        if ((flags & CMS_VERIFY_PARTIAL) != 0) {
```
according to coding style requirements

DDvO replied:
Maybe do the new signer info printing only in case `sk_CMS_SignerInfo_num(sinfos) > 1`,
since otherwise it is implicitly clear from the verification result.

DDvO commented on doc/man3/CMS_get0_SignerInfos.pod:
```diff
@@ -41,6 +45,16 @@ if not.
 B<signer>.
 
+CMS_SignerInfo_get0_signer_cert() returns the signer's certificate associated
+with B<si>. This pointer should not be freed.
```
```suggestion
with B<si>. This pointer should not be freed by the caller.
```

DDvO commented on doc/man3/CMS_get0_SignerInfos.pod:
```diff
@@ -78,6 +92,11 @@ Any error can be obtained from L<ERR_get_error(3)>
+=head1 HISTORY
+
+B<CMS_SignerInfo_get0_signer_cert> and
+B<CMS_SignerInfo_get_verification_result> were added in OpenSSL 3.6.
```
```suggestion
B<CMS_SignerInfo_get_verification_result> were added in OpenSSL 4.0.
```

DDvO commented on crypto/cms/cms_sd.c:
```diff
@@ -656,6 +656,11 @@ void CMS_SignerInfo_set1_signer_cert(CMS_SignerInfo *si, X509 *signer)
     si->signer = signer;
 }
 
+X509 *CMS_SignerInfo_get0_signer_cert(CMS_SignerInfo *si)
```
```suggestion
X509 *CMS_SignerInfo_get0_signer_cert(const CMS_SignerInfo *si)
```

DDvO commented on crypto/cms/cms_local.h:
```diff
@@ -103,6 +103,12 @@ struct CMS_SignerInfo_st {
     int omit_signing_time;
+    /* Remember which aspects have been verified */
+    int verify_result;
+    int verify_failure;
```
Aren't these two flags implied by the three following ones?
If so, remove each one that is not strictly needed,
and for each remaining one add a comment what exactly it means, also to explain why it is needed.

DDvO commented on crypto/cms/cms_smime.c:
```diff
@@ -303,6 +303,20 @@ static int cms_signerinfo_verify_cert(CMS_SignerInfo *si,
 
 }
 
+static int cms_signerinfos_remainig(STACK_OF(CMS_SignerInfo) *sinfos)
```
```suggestion
static int cms_signerinfos_remaining(const STACK_OF(CMS_SignerInfo) *sinfos)
```

DDvO commented on crypto/cms/cms_smime.c:
```diff
@@ -462,14 +500,57 @@ int CMS_verify(CMS_ContentInfo *cms, STACK_OF(X509) *certs,
+            continue;
+        if (!(flags & CMS_NO_SIGNER_CERT_VERIFY) && !si->cert_verified)
+            continue;
+        if (!(flags & CMS_NO_ATTR_VERIFY) && !si->attr_verified)
```
```suggestion
        if ((flags & CMS_NO_ATTR_VERIFY) == 0 && !si->attr_verified)
```

DDvO commented on crypto/cms/cms_smime.c:
```diff
@@ -462,14 +500,57 @@ int CMS_verify(CMS_ContentInfo *cms, STACK_OF(X509) *certs,
+            continue;
+        if (!(flags & CMS_NO_ATTR_VERIFY) && !si->attr_verified)
+            continue;
+        if (!(flags & CMS_NO_CONTENT_VERIFY) && !si->content_verified)
```
```suggestion
        if ((flags & CMS_NO_CONTENT_VERIFY) == 0 && !si->content_verified)
```

DDvO commented on crypto/cms/cms_smime.c:
```diff
@@ -462,14 +500,57 @@ int CMS_verify(CMS_ContentInfo *cms, STACK_OF(X509) *certs,
+        goto err;
+
+    /* Aggregate results per signer */
+    for (i = 0; i < sk_CMS_SignerInfo_num(sinfos); i++) {
```
```suggestion
    for (i = 0; i < scount; i++) {
```

DDvO commented on crypto/cms/cms_smime.c:
```diff
@@ -462,14 +500,57 @@ int CMS_verify(CMS_ContentInfo *cms, STACK_OF(X509) *certs,
+                break;
+            }
+        }
+    }
```
```suggestion
    if ((flags & CMS_VERIFY_PARTIAL) == 1)
        ret = cms_signerinfos_remainig(sinfos) > 0 /* One success is enough */
    else
        ret = cms_signerinfos_remainig(sinfos) == scount; /* All must be successful */
```

DDvO commented on doc/man3/CMS_verify.pod:
```diff
@@ -72,12 +73,16 @@ L<CMS_add1_cert(3)>, are used as untrusted CAs.
+The overall result of the function depends on whether B<CMS_VERIFY_PARTIAL> is
+set. When set, a single successfully verified signature is enough and
+L<CMS_get0_SignerInfos(3)> can be used to inspect each signature.
+Otherwise, the function is only successful if signatures verify correctly.
```
```suggestion
Otherwise, the function is only successful if all signatures verify sucessfully.
```

DDvO reviewed (changes requested):
Various small improvements

t8m commented:
I assume @DDvO just mis-clicked.

DDvO reviewed (changes requested):
Thanks for your very swift and cooperative reactions, making reviewer life easy.

A couple of follow-up comments/questions.
Part of which are a little further up in the thread but unfortunately folded away by github.com.

jluebbe commented:
I've applied the comment change requested [here](https://github.com/openssl/openssl/pull/27604#discussion_r2515760454).

I've not yet implemented [suppression of the detailed output](https://github.com/openssl/openssl/pull/27604#discussion_r2603245466) with `openssl cms`, as I'm not sure which alternative is preferred.

The main open question is how to store which aspects have been verified. I've [described why I chose the current approach](https://github.com/openssl/openssl/pull/27604#discussion_r2603417744), if you prefer the approach with only a boolean result and a bit field for aspect verification failures, I'll rewrite that part.

openssl-machine commented:
This PR is in a state where it requires action by @openssl/committers but the last update was 30 days ago

DDvO reviewed (changes requested):
Please remove the `verify_failure`, simplifying the code as suggested in the comments.

jluebbe commented:
I've force-pushed to integrate @DDvO's feedback (see diff below) and to rebase on master.

<details><summary>diff</summary>
<p>

```diff
diff --git a/apps/cms.c b/apps/cms.c
index 77f78b4fe5ea..d09d58171e29 100644
--- a/apps/cms.c
+++ b/apps/cms.c
@@ -1421,9 +1421,9 @@ int cms_main(int argc, char **argv)
                 X509_NAME_print_ex(bio_err, si_subject, 0, XN_FLAG_ONELINE);
                 BIO_printf(bio_err, "\n  Verification %s (cert: %s, attrs: %s, content: %s)\n",
                     CMS_SignerInfo_get_verification_result(si, CMS_VERIFY_RESULT) ? "successful" : "failed",
-                    CMS_SignerInfo_get_verification_result(si, CMS_VERIFY_CERT) ? "success" : "failure",
-                    CMS_SignerInfo_get_verification_result(si, CMS_VERIFY_ATTR) ? "success" : "failure",
-                    CMS_SignerInfo_get_verification_result(si, CMS_VERIFY_CONTENT) ? "success" : "failure");
+                    CMS_SignerInfo_get_verification_result(si, CMS_VERIFY_CERT) ? "success" : "failure or not verified",
+                    CMS_SignerInfo_get_verification_result(si, CMS_VERIFY_ATTR) ? "success" : "failure or not verified",
+                    CMS_SignerInfo_get_verification_result(si, CMS_VERIFY_CONTENT) ? "success" : "failure or not verified");
             }
         }
         if (signerfile != NULL) {
diff --git a/crypto/cms/cms_local.h b/crypto/cms/cms_local.h
index bbb398f154bb..e367de7a027f 100644
--- a/crypto/cms/cms_local.h
+++ b/crypto/cms/cms_local.h
@@ -105,9 +105,8 @@ struct CMS_SignerInfo_st {
     int omit_signing_time;
     /* Remember which aspects have been verified */
     int verify_result; /* for this SignerInfo, all attempted verification aspects succeeded */
-    int verify_failure; /* at least one verification aspect was attempted and failed */
     int cert_verified; /* the certificate was verified successfully */
-    int attr_verified; /* the certificate attributes were verified successfully */
+    int attr_verified; /* any given signed attributes were verified successfully */
     int content_verified; /* the signature over the content was verified successfully */
 };
 
diff --git a/crypto/cms/cms_sd.c b/crypto/cms/cms_sd.c
index adcdee1139b1..f49f27bb2336 100644
--- a/crypto/cms/cms_sd.c
+++ b/crypto/cms/cms_sd.c
@@ -1052,7 +1052,7 @@ int CMS_SignerInfo_get_verification_result(const CMS_SignerInfo *si, int type)
 {
     switch (type) {
     case CMS_VERIFY_RESULT:
-        return si->verify_result && !si->verify_failure;
+        return si->verify_result;
 
     case CMS_VERIFY_CERT:
         return si->cert_verified;
diff --git a/crypto/cms/cms_smime.c b/crypto/cms/cms_smime.c
index 9671d584d1ea..c11b5debe966 100644
--- a/crypto/cms/cms_smime.c
+++ b/crypto/cms/cms_smime.c
@@ -301,20 +301,6 @@ err:
     return r;
 }
 
-static int cms_signerinfos_remainig(const STACK_OF(CMS_SignerInfo) *sinfos)
-{
-    int i, remaining = 0;
-
-    for (i = 0; i < sk_CMS_SignerInfo_num(sinfos); i++) {
-        CMS_SignerInfo *si = sk_CMS_SignerInfo_value(sinfos, i);
-
-        if (si->verify_failure == 0)
-            remaining++;
-    }
-
-    return remaining;
-}
-
 /* This strongly overlaps with PKCS7_verify() */
 int CMS_verify(CMS_ContentInfo *cms, STACK_OF(X509) *certs,
     X509_STORE *store, BIO *dcont, BIO *out, unsigned int flags)
@@ -325,7 +311,7 @@ int CMS_verify(CMS_ContentInfo *cms, STACK_OF(X509) *certs,
     STACK_OF(X509_CRL) *crls = NULL;
     STACK_OF(X509) **si_chains = NULL;
     X509 *signer;
-    int i, scount = 0, ret = 0;
+    int i, n = 0, scount = 0, ret = 0;
     BIO *cmsbio = NULL, *tmpin = NULL, *tmpout = NULL;
     int cadesVerify = (flags & CMS_CADES) != 0;
     const CMS_CTX *ctx = ossl_cms_get0_cmsctx(cms);
@@ -354,8 +340,7 @@ int CMS_verify(CMS_ContentInfo *cms, STACK_OF(X509) *certs,
         if (signer != NULL)
             scount++;
         /* Reset verification results */
-        si->verify_result = 0;
-        si->verify_failure = 0;
+        si->verify_result = 1; /* so far, fine */;
         si->cert_verified = 0;
         si->attr_verified = 0;
         si->content_verified = 0;
@@ -395,36 +380,33 @@ int CMS_verify(CMS_ContentInfo *cms, STACK_OF(X509) *certs,
             if (!cms_signerinfo_verify_cert(si, store, untrusted, crls,
                     si_chains ? &si_chains[i] : NULL,
                     ctx)) {
-                si->verify_failure = 1;
+                si->verify_result = 0;
                 continue;
             }
             si->cert_verified = 1;
         }
     }
 
-    if (cms_signerinfos_remainig(sinfos) == 0)
-        goto err;
-
     /* Attempt to verify all SignerInfo signed attribute signatures */
 
     if ((flags & CMS_NO_ATTR_VERIFY) == 0 || cadesVerify) {
         for (i = 0; i < scount; i++) {
             si = sk_CMS_SignerInfo_value(sinfos, i);
-            if (si->verify_failure)
+            if (!si->verify_result)
                 continue;
             if (CMS_signed_get_attr_count(si) < 0) {
                 si->attr_verified = 1;
                 continue;
             }
             if (CMS_SignerInfo_verify(si) <= 0) {
-                si->verify_failure = 1;
+                si->verify_result = 0;
                 continue;
             }
             if (cadesVerify) {
                 STACK_OF(X509) *si_chain = si_chains ? si_chains[i] : NULL;
 
                 if (ossl_cms_check_signing_certs(si, si_chain) <= 0) {
-                    si->verify_failure = 1;
+                    si->verify_result = 0;
                     continue;
                 }
             }
@@ -432,9 +414,6 @@ int CMS_verify(CMS_ContentInfo *cms, STACK_OF(X509) *certs,
         }
     }
 
-    if (cms_signerinfos_remainig(sinfos) == 0)
-        goto err;
-
     /*
      * Performance optimization: if the content is a memory BIO then store
      * its contents in a temporary read only memory BIO. This avoids
@@ -496,42 +475,29 @@ int CMS_verify(CMS_ContentInfo *cms, STACK_OF(X509) *certs,
     if (!(flags & CMS_NO_CONTENT_VERIFY)) {
         for (i = 0; i < sk_CMS_SignerInfo_num(sinfos); i++) {
             si = sk_CMS_SignerInfo_value(sinfos, i);
-            if (si->verify_failure)
+            if (!si->verify_result)
                 continue;
             if (CMS_SignerInfo_verify_ex(si, cmsbio, tmpin) <= 0) {
                 ERR_raise(ERR_LIB_CMS, CMS_R_CONTENT_VERIFY_ERROR);
-                si->verify_failure = 1;
+                si->verify_result = 0;
                 continue;
             }
             si->content_verified = 1;
         }
     }
 
-    if (cms_signerinfos_remainig(sinfos) == 0)
-        goto err;
-
-    /* Aggregate results per signer */
-    for (i = 0; i < scount; i++) {
-        si = sk_CMS_SignerInfo_value(sinfos, i);
-
-        if (si->verify_failure)
-            continue;
-        if ((flags & CMS_NO_SIGNER_CERT_VERIFY) == 0 && !si->cert_verified)
-            continue;
-        if ((flags & CMS_NO_ATTR_VERIFY) == 0 && !si->attr_verified)
-            continue;
-        if ((flags & CMS_NO_CONTENT_VERIFY) == 0 && !si->content_verified)
-            continue;
-
-        si->verify_result = 1;
-    }
-
     /* Determine overall result */
+    for (i = 0; i < scount; i++) {
+        if (sk_CMS_SignerInfo_value(sinfos, i)->verify_result)
+            n++;
     if ((flags & CMS_VERIFY_PARTIAL) != 0)
-        ret = cms_signerinfos_remainig(sinfos) > 0; /* One success is enough */
+        ret = n > 0; /* One success is enough */
     else
-        ret = cms_signerinfos_remainig(sinfos) == scount; /* All must be successful */
+        ret = n == scount; /* All must be successful */
 err:
+    if (!ret)
+        for (i = 0; i < scount; i++)
+           sk_CMS_SignerInfo_value(sinfos, i)->verify_result = 0;
     if (!(flags & SMIME_BINARY) && dcont) {
         do_free_upto(cmsbio, tmpout);
         if (tmpin != dcont)
```

</p>
</details>

jluebbe commented:
Force-pushed to fix a [code style issue](https://github.com/openssl/openssl/actions/runs/21033434254/job/60474935938?pr=27604).

DDvO reviewed (approved):
LGTM

DDvO reviewed (changes requested):
The two nits just found by Copilot are relevant.

BTW, one could make the detailed result output in `cms.c` even more telling by making it depend on the flags `CMS_NO_SIGNER_CERT_VERIFY`, `CMS_NO_ATTR_VERIFY`, and `CMS_NO_CONTENT_VERIFY`. Just a thought in case you like it.

openssl-machine commented:
This PR is in a state where it requires action by @openssl/committers but the last update was 30 days ago

DDvO reviewed (changes requested):
Thanks for the update.
Just found one missing change 4.0 -> 4.1

openssl-machine commented:
This PR is in a state where it requires action by @openssl/committers but the last update was 30 days ago

jluebbe commented:
I've force-pushed to rebase on master. The changes are:

<details><summary>git range-diff origin/master 3bc712944d 5abbaa5ae0</summary>
<p>

```
1:  0cdf75db1c ! 1:  dc77e2591d Implement partial verification for CMS with multiple signatures
    @@ doc/man3/CMS_verify.pod: the content is detached I<detached_data> cannot be NULL
     +If any signing certificate cannot be located the operation fails unless
     +B<CMS_VERIFY_PARTIAL> is set, where one signing certificate can be sufficient.
      
    --Each signing certificate is chain verified using the I<smimesign> purpose and
    -+Each found signing certificate is chain verified using the I<smimesign> purpose and
    +-Each signing certificate is chain verified
    ++Each found signing certificate is chain verified
      using the trusted certificate store I<store> if supplied.
    - Any internal certificates in the message, which may have been added using
    - L<CMS_add1_cert(3)>, are used as untrusted CAs.
    + The purpose required in this verification is I<smimesign>
    + unless a different one (or B<X509_PURPOSE_DEFAULT_ANY>)
    +@@ doc/man3/CMS_verify.pod: L<CMS_add1_cert(3)>, are used as untrusted CAs.
      If CRL checking is enabled in I<store> and B<CMS_NOCRL> is not set,
      any internal CRLs, which may have been added using L<CMS_add1_crl(3)>,
      are used in addition to attempting to look them up in I<store>.
    @@ test/recipes/80-test_cms.t: my ($no_des, $no_dh, $no_dsa, $no_ec, $no_ec2m, $no_
      
      $no_rc2 = 1 if disabled("legacy");
      
    --plan tests => 34;
    -+plan tests => 35;
    +-plan tests => 37;
    ++plan tests => 38;
      
      ok(run(test(["pkcs7_test"])), "test pkcs7");
      
2:  6f99c8952c ! 2:  a2e7d1f085 Implement access to CMS verification results per signer
    @@ include/openssl/cms.h.in: void CMS_SignerInfo_get0_algs(CMS_SignerInfo *si, EVP_
      int CMS_SignerInfo_verify_content(CMS_SignerInfo *si, BIO *chain);
     
      ## util/libcrypto.num ##
    -@@ util/libcrypto.num: CMS_add1_signer                         ?	4_0_0	EXIST::FUNCTION:CMS
    - CMS_SignerInfo_get0_pkey_ctx            ?	4_0_0	EXIST::FUNCTION:CMS
    - CMS_SignerInfo_get0_md_ctx              ?	4_0_0	EXIST::FUNCTION:CMS
    - CMS_get0_SignerInfos                    ?	4_0_0	EXIST::FUNCTION:CMS
    +@@ util/libcrypto.num: OPENSSL_sk_set_cmp_thunks               5712	4_0_0	EXIST::FUNCTION:
    + ASN1_BIT_STRING_set1                    5713	4_0_0	EXIST::FUNCTION:
    + OSSL_ESS_check_signing_certs_ex         5714	4_0_0	EXIST::FUNCTION:
    + X509v3_delete_extension                 5715	4_0_0	EXIST::FUNCTION:
     +CMS_SignerInfo_get0_signer_cert         ?	4_1_0	EXIST::FUNCTION:CMS
    - CMS_SignerInfo_set1_signer_cert         ?	4_0_0	EXIST::FUNCTION:CMS
    - CMS_SignerInfo_get0_signer_id           ?	4_0_0	EXIST::FUNCTION:CMS
    - CMS_SignerInfo_cert_cmp                 ?	4_0_0	EXIST::FUNCTION:CMS
    -@@ util/libcrypto.num: CMS_SignerInfo_get0_signature           ?	4_0_0	EXIST::FUNCTION:CMS
    - CMS_SignerInfo_sign                     ?	4_0_0	EXIST::FUNCTION:CMS
    - CMS_SignerInfo_verify                   ?	4_0_0	EXIST::FUNCTION:CMS
    - CMS_SignerInfo_verify_content           ?	4_0_0	EXIST::FUNCTION:CMS
     +CMS_SignerInfo_get_verification_result  ?	4_1_0	EXIST::FUNCTION:CMS
    - CMS_SignedData_verify                   ?	4_0_0	EXIST::FUNCTION:CMS
    - CMS_add_smimecap                        ?	4_0_0	EXIST::FUNCTION:CMS
    - CMS_add_simple_smimecap                 ?	4_0_0	EXIST::FUNCTION:CMS
    + CTLOG_STORE_add0_log                    ?	4_1_0	EXIST::FUNCTION:CT
    + CRYPTO_atomic_load_ptr                  ?	4_1_0	EXIST::FUNCTION:
    + CRYPTO_atomic_store_ptr                 ?	4_1_0	EXIST::FUNCTION:
3:  3bc712944d = 3:  5abbaa5ae0 Update CHANGES.md for CMS_VERIFY_PARTIAL
```

</p>
</details>

jluebbe commented:
@DDvO @t8m Is there anything we could do to help with getting this merged? :)

DDvO commented:
> @DDvO @t8m Is there anything we could do to help with getting this merged? :)

I wanted to reconfirm my approval, 
but just noticed that it would be good to mention the addition of `CMS_VERIFY_PARTIAL` also in the HISTORY section of `CMS_verify.pod`, and likewise for `-verify_partial` in `openssl-cms.pod.in`.

jluebbe commented:
@DDvO I've added them to the respective HISTORY sections.

jluebbe commented:
Is there something we can do to help this along? This PR recently had it's first birthday. ;) 

We have a [PR in RAUC](https://github.com/rauc/rauc/pull/1802) which would start using this new functionality to allow new OTA bundle signing workflows. Easier CA key rollover is the first step for us, but M-of-N would be easy to add as well and has been requested a few times.

DDvO commented:
Ping @openssl/committers for 2nd review

bukka commented:
It's got conflicts so that needs to be resolved but FWIW otherwise the change makes sense to me.

bukka commented:
Also it should be run through clang-format if it wasn't already as the style check might be working now

bukka commented:
I removed the label just to see that.

jluebbe commented:
I've force-pushed to resolve the conflict with https://github.com/openssl/openssl/commit/d260311c53a0d656b880266f4c9c14c5f5da5bd8.

DDvO reviewed (dismissed):
> I've force-pushed to resolve the conflict with https://github.com/openssl/openssl/commit/d260311c53a0d656b880266f4c9c14c5f5da5bd8.

Thanks for fixing that trivial conflict.
Reconfirming my approval.

Good to see also that no coding style issue is reported anymore.

DDvO commented:
A new merge conflict has come up.

jluebbe commented:
I've force-pushed to resolve the conflict.

DDvO reviewed (dismissed):
The CI failure regarding quic is unrelated.

DDvO commented:
@openssl/committers this has been waiting for a 2nd review/approval for long.
My first approval was on Jan 15, after a through review and respective improvements.

beldmit commented:
Do I correctly remember that a signer certificate is not a mandatory part of signature structure? If yes, how does this code interact with verifier-provided certs?

jluebbe commented:
> Do I correctly remember that a signer certificate is not a mandatory part of signature structure?

Yes. A CMS signature contains separate sets of certificates and SignerInfos. The [set of certificates](https://datatracker.ietf.org/doc/html/rfc5652#section-10.2.3) is intended to be used for path building, but can be empty or even contain unrelated certs. It's not signed as part of the message.

> If yes, how does this code interact with verifier-provided certs?

The caller of [`CMS_verify`](https://docs.openssl.org/master/man3/CMS_verify/) can provide additional untrusted certificates via the `certs` parameter. Internally, both sources of certificates are simply combined in `STACK_OF(X509) *untrusted` and then used during verification of each individual signature.

This PR doesn't touch the handling of these certificates, because the current code already checks each signature indivually.

beldmit commented:
The function `CMS_SignerInfo_get0_signer_cert` you've introduced would work only with embedded certs, right?

jluebbe commented:
> The function `CMS_SignerInfo_get0_signer_cert` you've introduced would work only with embedded certs, right?

It works in either case: The `si->signer` pointer used by `CMS_SignerInfo_get0_signer_cert` is filled early in `CMS_verify`, by calling `CMS_set1_signers_certs` if any `SignerInfo`s are missing a certificate. `CMS_set1_signers_certs` loops over the `SignerInfo`s and tries to find matching certs first in those provided by the verfier and then in those contained the CMS message. Before continuing, `CMS_verify` checks if every `SignerInfo` has a certificate.

The existing code had to do this already, as the signer certificate verification (`cms_signerinfo_verify_cert`) also uses the `si->signer` pointer.

jluebbe commented:
@DDvO @beldmit Is there anything I could do to bring this forward? :)

beldmit commented:
Ping me next week when I return back to work

jluebbe commented:
Hi @beldmit, I've rebased to fix the conflict in `test/recipes/80-test_cms.t` regarding the number of tests.

beldmit reviewed (approved):
LGTM

jluebbe commented:
@DDvO What is the next step here?

DDvO reviewed (approved):
reconfirmed after the rebase.
Sorry for not noticing the request before.

DDvO commented:
For some reason, the "read to merge" label was not applied automatically.
Can this be merged now?

DDvO commented:
Maybe be "ready to merge" label was not applied automatically because
meanwhile yet another merge conflict with `CHANGES.md` arose?
Yet since it is trivial, this can be fixed during merging - ok?
Unless negative response arrives today, I'm going to do this tomorrow.

t8m commented:
Merged to the master branch. Thank you for your contribution.

jluebbe commented:
Thanks to everyone for getting this over the finish line!
